datalake/src/main/java/com/sequenceiq/datalake/flow/freeipa/upscale/FreeIpaUpscaleActions.java

```diff
@@ -0,0 +1,147 @@
+import static com.sequenceiq.datalake.flow.freeipa.upscale.FreeIpaUpscaleEvent.FREE_IPA_UPSCALE_IN_PROGRESS_EVENT;
+import static com.sequenceiq.datalake.flow.freeipa.upscale.FreeIpaUpscaleEvent.FREE_IPA_UPSCALE_SUCCESS_EVENT;
+
+import java.util.Map;
+import java.util.Optional;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.statemachine.StateContext;
+import org.springframework.statemachine.action.Action;
+
+import com.sequenceiq.datalake.flow.SdxContext;
+import com.sequenceiq.datalake.flow.SdxFailedEvent;
+import com.sequenceiq.datalake.flow.freeipa.upscale.event.FreeIpaUpscaleFailedEvent;
+import com.sequenceiq.datalake.flow.freeipa.upscale.event.FreeIpaUpscaleStartEvent;
+import com.sequenceiq.datalake.flow.freeipa.upscale.event.FreeIpaUpscaleSuccessEvent;
+import com.sequenceiq.datalake.flow.freeipa.upscale.event.FreeIpaUpscaleWaitRequest;
+import com.sequenceiq.datalake.metric.SdxMetricService;
+import com.sequenceiq.datalake.service.AbstractSdxAction;
+import com.sequenceiq.datalake.service.FreeipaService;
+import com.sequenceiq.flow.core.Flow;
+import com.sequenceiq.flow.core.FlowEvent;
+import com.sequenceiq.flow.core.FlowParameters;
+import com.sequenceiq.flow.core.FlowState;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.FormFactor;
+
+@Configuration
+public class FreeIpaUpscaleActions {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FreeIpaUpscaleActions.class);
+
+    @Inject
+    private FreeipaService freeipaService;
+
+    @Inject
+    private SdxMetricService metricService;
+
+    @Bean(name = "FREE_IPA_UPSCALE_START_STATE")
+    public Action<?, ?> freeIpaUpscale() {
+        return new AbstractSdxAction<>(FreeIpaUpscaleStartEvent.class) {
+            @Override
+            protected SdxContext createFlowContext(FlowParameters flowParameters, StateContext<FlowState, FlowEvent> stateContext,
+                    FreeIpaUpscaleStartEvent payload) {
+                return SdxContext.from(flowParameters, payload);
+            }
+
+            @Override
+            protected void doExecute(SdxContext context, FreeIpaUpscaleStartEvent payload, Map<Object, Object> variables) {
+
+                if (FormFactor.HA.getInstanceCount() > freeipaService.getNodeCount(payload.getEnvCrn())) {
+                    LOGGER.info("FreeIpa Upscale has been started for {}", payload.getResourceId());
+                    String operationId = freeipaService.upscale(payload.getEnvCrn());
+                    payload.setOperationId(operationId);
+                    sendEvent(context, FREE_IPA_UPSCALE_IN_PROGRESS_EVENT.event(), payload);
+                } else {
+                    LOGGER.info("FreeIpa Upscale being skipped for {}", payload.getResourceId());
```

Comment:
Please update the log message with the reason why the upscale is skipped. This makes it clear.